### ISSUE TITLE
Can you explain how you would write a driver for Linux?

### ISSUE THREAD

**Issue by PranabNandy** [PranabNandy/Linux-Kernel-Internals#3]
🤓Me: Okay..the first step is really getting to know the device we are dealing with. That means diving into the datasheets and making sure I understand the communication protocols, whether it is I2C, SPI, or UART. It is all about getting that solid understanding from the start.

🥸Interviewer: That makes sense. What would you do next?

🤓Me: After that, I would focus on the kernel module itself. **`In Linux, most drivers are kernel modules,`** so I would start with the basics—setting up the module’s initialization and cleanup functions. It is kind of like doing a “Hello, World!” but for the kernel.

🥸Interviewer: And how would you handle the interaction with the device?

🤔Me: That is where the device file ( `/dev/xyz_device` )system comes into play. Linux treats hardware devices like files, so **`I would create a device file to let user-space applications talk to the driver.`** Implementing functions like open, read, write, and **ioctl** would be key here. Also, I will ensure that the necessary entries in the DTS are correctly configured to match the driver.

🤨Interviewer: How do you deal with hardware interrupts in your driver (The interrupt generated by hardware device)?

😵‍💫Me: Handling interrupts efficiently is really important. I would make sure the driver registers the right interrupt handlers and manages them so we do not end up with latency issues or miss any important events.

**Here is 2 things:**
- Driver will register the right interrupt handler
- Driver will manage that interrupt handler

🧐Interviewer: Sounds good. Anything tricky that you keep an eye on?

😵‍💫Me: Definitely. Memory management is a big one. Kernel memory is not like user space, so I would be r**eally careful to allocate and free memory properly to avoid any leaks.** Also, dealing with concurrency is super important—**using spinlocks or semaphores** to protect shared resources helps prevent race conditions.

😉Interviewer: And what about debugging?

😅Me: Debugging in the kernel can be a bit tricky, but tools **`like dmesg, printk, and gdb`** are really helpful. I would make sure to test the driver under different conditions to catch any issues early on.

Interviewer: Hmmm ...anything else ?

😌Me: Adding a new device is not just about making it work—it is about making sure it fits well with the Linux kernel and plays nicely with other drivers. Plus, **I would make sure the code follows the open-source licensing rules since Linux uses the GPL license.**

😇Interviewer: Yea right..
 Good explanation, lets move forward.

**Comment by PranabNandy** [comment]
# ioctrl
1.Define ioctl first

```cpp
#define "ioctl name" __IOX("magic number","command number","argument type")
```
where IOX can be :

- “IO“: an ioctl with no parameters
- “IOW“: an ioctl with write parameters (copy_from_user)
- “IOR“: an ioctl with read parameters (copy_to_user)
- “IOWR“: an ioctl with both write and read parameters

- The Magic Number is a unique number or character that will differentiate our set of ioctl calls from the other ioctl calls. some times the major number for the device is used here.

- Command Number is the number that is assigned to the ioctl. This is used to differentiate the commands from one another.

- The last is the type of data.
```cpp
#include <linux/ioctl.h>
#define WR_VALUE _IOW('a','a',int32_t*)
#define RD_VALUE _IOR('a','b',int32_t*)
```
2.
```cpp

/*
** This function will be called when we write IOCTL on the Device file
*/
static long etx_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
{
         switch(cmd) {
                case WR_VALUE:
                        if( copy_from_user(&value ,(int32_t*) arg, sizeof(value)) )
                        {
                                pr_err("Data Write : Err!\n");
                        }
                        pr_info("Value = %d\n", value);
                        break;
                case RD_VALUE:
                        if( copy_to_user((int32_t*) arg, &value, sizeof(value)) )
                        {
                                pr_err("Data Read : Err!\n");
                        }
                        break;
                default:
                        pr_info("Default\n");
                        break;
        }
        return 0;
}
/*
** File operation sturcture
*/
static struct file_operations fops =
{
        .owner          = THIS_MODULE,
        .read           = etx_read,
        .write          = etx_write,
        .open           = etx_open,
        .unlocked_ioctl = etx_ioctl,
        .release        = etx_release,
};
```

**Comment by PranabNandy** [comment]
Static vs. Dynamic Linking and Loading 

Whether you’re just starting out or you’re a seasoned pro, it’s always good to refresh our knowledge on these fundamental concepts. So, let’s break it down!

Static Linking:
🔗 Imagine you’re baking a cake, and you mix all your ingredients (code libraries) directly into the batter (your program). Once baked, everything is a single, solid cake. This is what static linking does – it combines all the necessary libraries with your code into one executable file at compile time. The result? Fast execution because everything’s already in place, but it comes at the cost of a larger file size. 📦

Dynamic Linking:
🔗 Now, picture making a cake but keeping the frosting (code libraries) separate until you serve it. This is dynamic linking – your program and its libraries remain separate entities. The linking happens at runtime, meaning the program will call the necessary libraries when needed. This keeps your executable smaller and allows updates to libraries without recompiling the entire program, but it can be a bit slower due to the additional steps during execution. ⚡

Loading:
🔄 Both static and dynamic linking involve loading. For static, the loading happens once when the program is started. For dynamic, loading can happen multiple times during the program’s execution, whenever a library is needed.

Adding Sugar to the Frosting:
🍰 Think about dynamic linking like deciding to add more sugar to the frosting just before serving the cake. If you realize your frosting (code library) needs a tweak or an update, you can easily mix in the extra sugar (update the library) without having to rebake the entire cake (recompile the program). This flexibility is a huge advantage in environments where updates and changes are frequent.

When to Use What?
🛠️ Use *static linking* when you need speed and don’t mind the larger file size, like in embedded systems where you want everything tightly packed and optimized for performance.

🛠️ Use dynamic linking when you want flexibility and ease of updates, perfect for applications where libraries might change or get updated frequently.

In the end, the choice between static and dynamic linking and loading depends on your specific needs and constraints. Understanding these concepts helps us make informed decisions in our software projects.

Happy coding 🥸

**Comment by PranabNandy** [comment]
🤖 DMA and👨‍🍳 Cooking at Home 🌟

🤔What is DMA?

Direct Memory Access (DMA) is a feature that allows certain hardware subsystems to access main system memory (RAM) independently of the central processing unit (CPU). In simpler terms, DMA can transfer data between memory and peripherals without burdening the CPU.

Imagine you are cooking a meal🤤 with your spouse. You are busy at the stove, managing multiple dishes, while your spouse🐒 helps by fetching ingredients from the fridge and the pantry. If you had to stop cooking every time to get ingredients yourself, it would slow down the whole process🐢.

In this analogy, imgine you are the CPU, the kitchen is the memory, and your spouse is the DMA controller. Your spouse handles the task of bringing ingredients, allowing you to focus on cooking without interruptions. So you need not to pull out onions potatos and wash hands and go back and start cooking again. Similar is with DMA, cpu need not to worry about waiting for the data and doing data read writes and data flushing itself while executing something more important🥸

So what does DMA offers 🧐 :
1. *Efficiency and Speed*: DMA significantly speeds up data transfers. By offloading the data transfer tasks to the DMA controller, the CPU is free to execute other instructions, leading to more efficient system performance.

2. *Reduced CPU Load*: In embedded systems, the CPU often has to manage multiple tasks simultaneously. DMA reduces the CPU's workload by handling the bulk of data movement, allowing the CPU to focus on critical real-time tasks.

3. *Low Power Consumption*: Embedded systems, especially those in battery-powered devices, benefit from DMA's ability to perform data transfers with minimal CPU intervention. This helps in conserving power and extending battery life.

4. *Improved Data Handling*: DMA is particularly useful for handling large blocks of data, such as streaming audio or video, where timely and efficient data transfer is crucial. It ensures data integrity and smooth operation of multimedia applications.

In summary, DMA is like having a dedicated helper in the kitchen who brings you ingredients while you cook. It enhances efficiency, performance, and power management in embedded systems. Understanding and utilizing DMA can make a significant difference in the design and functionality of embedded applications.

**Comment by PranabNandy** [comment]
🧐𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: Ok and have you ever worked on ARM Cortex-M0 processors before? 
 
😎𝐌𝐞: Yes, I'm familiar with M0 processor. I have worked on it under various projects that i have worked on. 
 
🧐𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: I see, so have you heard about the NVIC in ARM Cortex-M0 processors? 
 
😎𝐌𝐞: Yes Its an interrupt controller 
 
🧐𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: Do you know what it stands for ? 
 
🥲𝐌𝐞: Well the last part "IC" stands for Interrupt controller😬 
 
🫡𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: I see 😐, so what its used for in M0 ? 
 
🤓You: NVIC is a crucial component in ARM Cortex-M0 processors. It serves as the traffic manager for interrupts, ensuring efficient handling of various events. 
 
🧐𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: Can you elaborate more on its role in M0 processor? 
 
🤔𝐌𝐞: NVIC's nested and prioritized structure allows for the seamless handling of multiple interrupt sources. It ensures that critical tasks are addressed promptly, contributing to the overall efficiency of the system. 
 
🧐𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: And how does the NVIC contribute to energy efficiency in embedded systems? 
 
🥸𝐌𝐞: The NVIC helps in energy efficiency by allowing the processor to enter low-power states when not actively processing interrupts. This dynamic power management ensures that the system conserves energy without compromising responsiveness. 
 
🧐𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: Correct and how about real-time responsiveness? How does the NVIC handle high-priority interrupts? 
 
🤔𝐌𝐞: Umm.. NVIC's nested structure enables immediate handling of high-priority interrupts, minimizing latency and ensuring that time-sensitive tasks are executed with precision. This makes it ideal for applications requiring real-time responsiveness, such as sensor networks and motor control systems. 
 
🧐𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: Can we say it has good scalability? or... umm how the NVIC adapts to different project sizes? 
 
🤔𝐌𝐞: NVIC is has good scalability. Whether you're working on a small-scale project or a complex embedded system, the NVIC adapts seamlessly, allowing developers to allocate priorities based on the criticality of each interrupt source. This flexibility empowers developers to fine-tune system behavior to meet the unique demands of their applications. 
 
😈𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: Can you name some more Interrupt controllers ?

**Comment by PranabNandy** [comment]
🧐𝗜𝗻𝘁𝗲𝗿𝘃𝗶𝗲𝘄𝗲𝗿: "Which paging technique is used in RTOS?" 
 
🥸𝗠𝗲: "In RTOS, demand paging is commonly used." 
 
🧐𝗜𝗻𝘁𝗲𝗿𝘃𝗶𝗲𝘄𝗲𝗿: "Could you explain what demand paging is?" 
 
😎𝗠𝗲: "Certainly! Demand paging is a memory management technique where pages are loaded into memory only when they are required by the executing program. It helps optimize memory usage by bringing in only the necessary pages, reducing overhead and improving overall system performance." 
 
🧐𝗜𝗻𝘁𝗲𝗿𝘃𝗶𝗲𝘄𝗲𝗿: "Interesting! What is the primary motive of RTOS then?" 
 
😎𝗠𝗲: "The primary goal of RTOS is to ensure quick response times and deterministic behavior for real-time applications."
 
😈𝗜𝗻𝘁𝗲𝗿𝘃𝗶𝗲𝘄𝗲𝗿: "Considering that, can demand paging be used in RTOS?" 
 
🥲𝗠𝗲: "You're absolutely correct. Given that the primary objective of RTOS is rapid response times, demand paging, with its dynamic loading of pages from disk to memory, might not be suitable." 
 
😈𝗜𝗻𝘁𝗲𝗿𝘃𝗶𝗲𝘄𝗲𝗿: So what do we use in RTOS? 
 
𝗠𝗲: 😶

**Comment by PranabNandy** [comment]
🤓Preparing for a telephonic Interview ? Here are some trivial yet crucial questions you might encounter:

⁉️ Difference between a hard link and a soft link in Linux? 
✅️ A hard link points directly to the file's inode, while a soft link points to the file's path.

⁉️What is a shell in Linux? 
✅️ The shell is a command-line interpreter facilitating user-OS interaction and command execution.

⁉️How to check the memory usage of a Linux system? 
✅️Use the "free" command to display memory usage, including total, used, free memory, and swap usage.

⁉️How to find the IP address of a Linux system? 
✅️Utilize the "ifconfig" command to display network interface information, including IP addresses.

⁉️How to find the list of running processes in Linux? 
✅️Employ the "ps" command with options like "-ef" or "-aux" to display running processes.

⁉️How to kill a process in Linux? ✅️Terminate a process using the "kill" command, requiring the process ID obtained from "ps.

⁉️"Purpose of the "chmod" command in Linux?
✅️"chmod" changes file and directory permissions, granting or revoking read, write, and execute permissions.

⁉️How to search for a file in Linux? 
✅️Use the "find" command to search for files based on various criteria like name, size, or modification time.

⁉️What is the purpose of the "cron" daemon in Linux? 
✅️The "cron" daemon is a time-based job scheduler in Linux, allowing users to automate command execution at specified intervals.

⁉️How do you schedule a cron job in Linux? 
✅️To schedule a cron job, use the "crontab" command. Running "crontab -e" opens the user's crontab file to specify the command or script and schedule.

⁉️Difference between SSH and SSL? ✅️SSH (Secure Shell) is for secure remote access and system management, while SSL (Secure Sockets Layer) establishes encrypted connections for secure web communication (HTTPS).

⁉️How to check network connectivity in Linux? 
✅️Use the "ping" command to check connectivity between your machine and a remote host by sending ICMP echo requests and displaying response time.

🤓Do share your experience for telephonic interview questions and answers that you might have stumbled upon.

**Comment by PranabNandy** [comment]
🤔Imagine your computer is like a house🏠, and the BIOS (Basic Input/Output System) is like the lock🔒on the front door.

Just as the lock controls access to your house, the BIOS controls access🚫 to your computer's hardware when it starts up.

Now, think of Coreboot as a customizable, open-source security system for your house. Instead of relying on a standard lock, Coreboot allows you to design your own security measures 🔐✨️ tailored to your specific needs. You can choose the types of locks, alarms, and access controls you want to implement, giving you more control and transparency over your home's security💃💃.

So you have a traditional house with a standard lock on the front door (BIOS). However, you're concerned about security vulnerabilities and limitations with the standard lock. So, you decide to install Coreboot, which allows you to customize your home's security system.

😮‍💨With BIOS (Standard Lock):

You have limited control over the lock's features and settings.If there's a security flaw in the lock, you have to wait for the manufacturer to release a fix. You can't easily modify or upgrade the lock to better suit your needs.

🤓With Coreboot (Custom Security System):You have full control over the security features of your home.If there's a security issue, you can directly address it or rely on the active community to provide solutions. You can customize and upgrade your security system at any time, adding new features or improving existing ones.

So, Coreboot empowers😎 you to take control of your computer's firmware, just like a customizable security system gives you more control over your home's security. It's all about flexibility, transparency, and putting the power back in your hands.

**Comment by PranabNandy** [comment]
🤤 Imagine you're a head chef (representing a process) in a bustling kitchen, tasked with creating a masterpiece dish. You have a limited number of ingredients (resources) at your disposal, and if you attempt to whip up the entire dish from start to finish solo, you'll end up spending a lot of time juggling tasks😰.

But fear not! 🌟 You've hired a team of eager intern chefs🧑‍🍳🧑‍🍳🧑‍🍳 (representing threads) who may not know the recipe from start to finish, but they excel at handling individual tasks like frying, seasoning, or marinating. Just like threads in an operating system, these interns can work simultaneously and independently, each focusing on a specific aspect of the dish.

Now, here's where the magic happens! 🎩✨ Instead of hoarding all the ingredients to yourself, you share them with your intern chefs, allowing them to prepare their assigned tasks in parallel. While one intern is sautéing onions, another can be grilling the protein, and yet another can be prepping the garnishes.

By delegating tasks to your team of intern chefs (threads), you optimize efficiency and minimize downtime. Plus, if one task requires a bit more time or resources, it doesn't stall the entire cooking process. Just like in operating systems, where threads can be managed and executed concurrently, this culinary analogy illustrates the power of parallelism and resource sharing.

So, embrace the kitchen chaos and let your intern chefs (threads) work their culinary magic, all under the guidance of the head chef (process). Bon appétit! 🍽️

**Comment by PranabNandy** [comment]
🥹𝗗𝗮𝘁𝗮 𝘀𝗲𝗻𝘁 𝗮𝗻𝗱 𝗲𝘅𝗽𝗲𝗰𝘁𝗲𝗱 𝗻𝗼𝘁 𝗺𝗮𝘁𝗰𝗵𝗶𝗻𝗴 ?
Have you ever encountered memory alignment issues or wanted to reduce the memory footprint of your C/C++ structs? 🧐
The __𝐚𝐭𝐭𝐫𝐢𝐛𝐮𝐭𝐞__((𝐩𝐚𝐜𝐤𝐞𝐝)) attribute might just be the tool you need!
In C and C++, structures are typically padded with extra bytes to ensure proper alignment, which can lead to wasted memory space, especially in embedded systems or when working with network protocols. But fear not!
The __𝐚𝐭𝐭𝐫𝐢𝐛𝐮𝐭𝐞__((𝐩𝐚𝐜𝐤𝐞𝐝)) attribute comes to the rescue by instructing the compiler to pack structure members without any padding. Here's a quick example:
[hashtag#𝐢𝐧𝐜𝐥𝐮𝐝𝐞](https://www.linkedin.com/feed/hashtag/?keywords=include&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7175040270437158912) <𝐬𝐭𝐝𝐢𝐨.𝐡>
𝐬𝐭𝐫𝐮𝐜𝐭 𝐄𝐱𝐚𝐦𝐩𝐥𝐞 {
 𝐜𝐡𝐚𝐫 𝐚;
 𝐢𝐧𝐭 𝐛;
 𝐜𝐡𝐚𝐫 𝐜;
} __𝐚𝐭𝐭𝐫𝐢𝐛𝐮𝐭𝐞__((𝐩𝐚𝐜𝐤𝐞𝐝));
𝐢𝐧𝐭 𝐦𝐚𝐢𝐧() {
 𝐩𝐫𝐢𝐧𝐭𝐟("𝐒𝐢𝐳𝐞 𝐨𝐟 𝐬𝐭𝐫𝐮𝐜𝐭 𝐄𝐱𝐚𝐦𝐩𝐥𝐞: %𝐳𝐮 𝐛𝐲𝐭𝐞𝐬\𝐧", 𝐬𝐢𝐳𝐞𝐨𝐟(𝐬𝐭𝐫𝐮𝐜𝐭 𝐄𝐱𝐚𝐦𝐩𝐥𝐞));
 𝐫𝐞𝐭𝐮𝐫𝐧 𝟎;
}
🤓By applying __attribute__((packed)) to the Example struct, we eliminate any padding between its members, resulting in a more compact memory layout. However, it's crucial to note that using this attribute may affect performance due to potential alignment issues, so it's best suited for specific use cases where memory optimization is paramount.
😎Next time you're optimizing memory usage or working on a project with strict memory constraints, consider harnessing the power of __attribute__((packed)) to streamline your data structures and enhance efficiency!

**Comment by PranabNandy** [comment]
Ever heard about IP addresses and MAC addresses in networking systems ? what are these two terms ?

😆We can say mac_addr is pet name or ghar ka naam e.g. munna and ip_addr is your social media tag/username e.g princess_angel

So the social media would know you by your social media name or IP_ADDR.
And in your home everybody knows you via your pet name or mac_addr.

So in terms of networking let's say you have a home network with multiple devices connected to a Wi-Fi router. 

Say your smartphone has an IP address of [192.xxx.x.xxx](http://192.xxx.x.xxx/) assigned to it by the router.

And Your laptop has an IP address of [192.xxx.x.xxx](http://192.xxx.x.xxx/) assigned to it by the router.

And these IP addresses are used for devices to communicate with each other over the network and to access the internet.

And when we talk about MAC Address 

Your smartphone has a MAC address like 00:1A:xx:xx:xx:xx.
Your laptop has a MAC address like 11:22:xx:xx:xx:xx.

MAC addresses are used by the router to direct data specifically to each device within the local network(ghar).When data packets are sent between devices on the local network, the router uses the MAC address to determine which device to send the data to.

So, while IP addresses help devices communicate across networks and the internet, MAC addresses are used within a local network to ensure that data is sent to the correct device.

Both IP addresses and MAC addresses work together to facilitate communication and data transmission in computer networks.

**Comment by PranabNandy** [comment]
As C developers, we heavily rely on this function, and it's often a hot topic in interviews.

One of the most commonly asked questions is: "Can you describe and demonstrate the use of `malloc()`?"🧐

So, let's dive into a few important points about 💁🏻‍♂️`malloc()`:

1. 𝗗𝘆𝗻𝗮𝗺𝗶𝗰 𝗠𝗲𝗺𝗼𝗿𝘆 𝗔𝗹𝗹𝗼𝗰𝗮𝘁𝗶𝗼𝗻: `malloc()` in C allows dynamic allocation of memory during runtime, enabling programs to adapt to varying memory needs.

2. 𝗛𝗲𝗮𝗽 𝗠𝗲𝗺𝗼𝗿𝘆: Memory allocated by `malloc()` comes from the heap, a region of the process's virtual memory space, providing flexibility in memory management.

3. 𝗦𝘆𝗻𝘁𝗮𝘅: The syntax of `malloc()` is `void* malloc(size_t size)`, where `size` specifies the number of bytes to allocate.

4. 𝗥𝗲𝘁𝘂𝗿𝗻 𝗩𝗮𝗹𝘂𝗲: It returns a pointer to the beginning of the allocated memory block or NULL if the allocation fails.

5. 𝗨𝘀𝗮𝗴𝗲: `malloc()` is commonly used for allocating memory for arrays, structs, and other data structures whose size may vary at runtime.

6. 𝗠𝗲𝗺𝗼𝗿𝘆 𝗟𝗲𝗮𝗸 𝗣𝗿𝗲𝘃𝗲𝗻𝘁𝗶𝗼𝗻: Proper use of `malloc()` helps prevent memory leaks by allocating memory dynamically and freeing it when no longer needed using `free()`.

7. 𝗘𝗿𝗿𝗼𝗿 𝗛𝗮𝗻𝗱𝗹𝗶𝗻𝗴: It's essential to check if `malloc()` returns NULL to handle failed memory allocation gracefully.

8. 𝗧𝘆𝗽𝗲𝗰𝗮𝘀𝘁𝗶𝗻𝗴: Although not required in C, it's a good practice to typecast the returned pointer from `malloc()` to the appropriate type.

9. 𝗦𝗶𝘇𝗲 𝗖𝗮𝗹𝗰𝘂𝗹𝗮𝘁𝗶𝗼𝗻: The size argument in `malloc()` should be calculated based on the size of the data type being allocated and the number of elements required.

10. 𝗠𝗲𝗺𝗼𝗿𝘆 𝗢𝗽𝘁𝗶𝗺𝗶𝘇𝗮𝘁𝗶𝗼𝗻: By dynamically allocating memory only when needed, `malloc()` optimizes memory usage and improves program efficiency.

Understanding and mastering `malloc()` is crucial for efficient memory management in C programming, enabling developers to build robust and scalable applications.

**Comment by PranabNandy** [comment]
1) What is a Makefile, and what role does it play in software development?🧐

-> Makefiles automate the build process in software development by specifying rules for compiling, linking, and executing code files.

🤓So say if we have very huge number of files and directories that need to be compiled and they all have dependencies on certain header files and other files🥸, so compiling them one by one will be next to impossible that's when makefile comes into action.

2) How are rules defined in a Makefile, and what are their components?🤨

-> Rules consist of targets, dependencies, and commands. For example:

target: dependency1 dependency2
 { your command }

3) Why is dependency management important in Makefiles, and how does it optimize the build process?

-> Dependency management minimizes unnecessary recompilation by tracking file relationships. 

Example:
target: dependency
 {command}

4) What are implicit rules in Makefiles, and when would you use them?🧐 Provide an example.

->Implicit rules are pre-defined templates for common tasks, simplifying Makefile maintenance.

Example:
%.o: %.c
 $(CC) -c $< -o $@

5) How can variables be utilized in Makefiles, and what advantages do they offer?🥸

Variables enhance flexibility, consistency, and readability.

Example:

CC = gcc
CFLAGS = -Wall -O2
SRC = main.c utils.c
OBJ = $(SRC:.c=.o)
TARGET = my_program


😎Mastering these Makefile interview questions will not only showcase your expertise but also set you apart as a skilled software developer ready to tackle complex build processes with ease. 

💪So, are you prepared to tackle Makefile questions in your next interview?

**Comment by PranabNandy** [comment]
🥳 𝗦𝗶𝗺𝗽𝗹𝗲 𝗦𝘆𝘀𝘁𝗲𝗺 𝗱𝗲𝘀𝗶𝗴𝗻 𝗶𝗻 𝗘𝗺𝗯𝗲𝗱𝗱𝗲𝗱 𝗦𝗼𝗳𝘁𝘄𝗮𝗿𝗲 𝗜𝗻𝘁𝗲𝗿𝘃𝗶𝗲𝘄🤓💰

🥸𝗜𝗻𝘁𝗲𝗿𝘃𝗶𝗲𝘄𝗲𝗿 (🅸): Shashank, Can you explain the use of interrupts? Why do we need interrupts? how do they work?

🤓🅼🅴: Sure! Interrupt mechanism allows processor to respond to ext. events without constantly polling status. When interrupt occurs, processor stops executing current code, saves context, and jumps to ISR to handle the event.
After ISR is complete, processor resumes its previous operation.

🥸🅸: Hmmm.. correct. Now, let's say you're working on an embedded system with button presses and sensor readings as interrupt sources. What exactly you need to do to prioritize and manage these interrupts?

🤔🅼🅴: Okay let me explain step wise what I need to do:

𝟭.𝗖𝗼𝗻𝗳𝗶𝗴𝘂𝗿𝗲 𝗜𝗻𝘁𝗲𝗿𝗿𝘂𝗽𝘁 𝗦𝗼𝘂𝗿𝗰𝗲𝘀: I can set up INT0 to trigger on a falling edge (for button presses) and INT1 to trigger on a rising edge (for sensor readings).

𝟮.𝗘𝗻𝗮𝗯𝗹𝗲 𝗜𝗻𝘁𝗲𝗿𝗿𝘂𝗽𝘁𝘀: I would enable the respective interrupts using EIMSK (External Interrupt Mask Register).

𝟯. 𝗗𝗲𝗳𝗶𝗻𝗲 𝗜𝗦𝗥𝘀: I'd write separate ISRs for each interrupt source. In the ISRs, I'd handle the specific tasks related to the interrupt source.

In the main loop, I continuously check the interrupt flags associated with each interrupt source. If an interrupt flag is set, I'd execute the corresponding ISR to handle the event and reset the flag afterwards.

🥸🅸: Cool! Please write a sample code?

🥹🅼🅴: Ok! Here's an ex. code in C:

// 𝐃𝐞𝐟𝐢𝐧𝐞 𝐟𝐥𝐚𝐠𝐬 𝐭𝐨 𝐭𝐫𝐚𝐜𝐤 𝐢𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭 𝐬𝐭𝐚𝐭𝐮𝐬
𝐯𝐨𝐥𝐚𝐭𝐢𝐥𝐞 𝐮𝐢𝐧𝐭8_𝐭 𝐛𝐮𝐭𝐭𝐨𝐧𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐅𝐥𝐚𝐠 = 0; 
𝐯𝐨𝐥𝐚𝐭𝐢𝐥𝐞 𝐮𝐢𝐧𝐭8_𝐭 𝐬𝐞𝐧𝐬𝐨𝐫𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐅𝐥𝐚𝐠 = 0;

𝐯𝐨𝐢𝐝 𝐞𝐧𝐚𝐛𝐥𝐞𝐆𝐥𝐨𝐛𝐚𝐥𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐬()
{
 𝐬𝐞𝐢(); 
}

// 𝐈𝐒𝐑 𝐟𝐨𝐫 𝐛𝐮𝐭𝐭𝐨𝐧 𝐩𝐫𝐞𝐬𝐬𝐞𝐬
𝐈𝐒𝐑(𝐈𝐍𝐓0_𝐯𝐞𝐜𝐭)
{
 𝐛𝐮𝐭𝐭𝐨𝐧𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐅𝐥𝐚𝐠 = 1;
}
// ISR for sensor
𝐈𝐒𝐑(𝐈𝐍𝐓1_𝐯𝐞𝐜𝐭)
{
 𝐬𝐞𝐧𝐬𝐨𝐫𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐅𝐥𝐚𝐠 = 1;
}
𝐢𝐧𝐭 𝐦𝐚𝐢𝐧(𝐯𝐨𝐢𝐝)
{
 // 𝐈𝐧𝐢𝐭𝐢𝐚𝐥𝐢𝐳𝐞 𝐆𝐏𝐈𝐎 𝐚𝐧𝐝 𝐜𝐨𝐧𝐟𝐢𝐠𝐮𝐫𝐞 𝐢𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐬
 // (𝐂𝐨𝐧𝐟𝐢𝐠𝐮𝐫𝐞 𝐈𝐍𝐓0 𝐚𝐧𝐝 𝐈𝐍𝐓1 𝐚𝐬 𝐦𝐞𝐧𝐭𝐢𝐨𝐧𝐞𝐝 𝐞𝐚𝐫𝐥𝐢𝐞𝐫)

 𝐄𝐈𝐂𝐑𝐀 |= (1 << 𝐈𝐒𝐂01);
 // Set falling edge trigger for INT0
 
𝐄𝐈𝐂𝐑𝐀 |= (1 << 𝐈𝐒𝐂11) | (1 << 𝐈𝐒𝐂10); 
// Set Rising edge trigger for INT1
 
𝐄𝐈𝐌𝐒𝐊 |= (1 << 𝐈𝐍𝐓0) | (1 << 𝐈𝐍𝐓1);
// Enabling INT1 & INT2

 𝐞𝐧𝐚𝐛𝐥𝐞𝐆𝐥𝐨𝐛𝐚𝐥𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐬();

 𝐰𝐡𝐢𝐥𝐞 (1) // GIC handles this
 {
 𝐢𝐟 (𝐛𝐮𝐭𝐭𝐨𝐧𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐅𝐥𝐚𝐠)
 {
 //ʜᴀɴᴅʟᴇ button press interrupt & clear it
 𝐛𝐮𝐭𝐭𝐨𝐧𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐅𝐥𝐚𝐠 = 0;
 }
 𝐢𝐟 (𝐬𝐞𝐧𝐬𝐨𝐫𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐅𝐥𝐚𝐠)
 {
 //ʜᴀɴᴅʟᴇ ꜱᴇɴꜱᴏʀ ʀᴇᴀᴅɪɴɢ ɪɴᴛᴇʀʀ. & clear it
 𝐬𝐞𝐧𝐬𝐨𝐫𝐈𝐧𝐭𝐞𝐫𝐫𝐮𝐩𝐭𝐅𝐥𝐚𝐠 = 0;
 }
 }
 𝐫𝐞𝐭𝐮𝐫𝐧 0;
}

🅸: Can you explain why using "volatile" ?

**Comment by PranabNandy** [comment]
🥸𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: What Linux flavour you are most used to ? 
 
🤓𝐌𝐞: Ubuntu 🐧 mostly 
 
🥸𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: So you use Ubuntu ? Do you know about its boot flow ? Can you explain ? 
 
🤓𝐌𝐞: Sure (grabs Marker and hops 🚀 infront of white board) 
 
𝟏. 𝐁𝐈𝐎𝐒/𝐔𝐄𝐅𝐈: It all starts with the Basic Input/Output System (BIOS) or Unified Extensible Firmware Interface (UEFI) initializing hardware and loading the bootloader. 
 
𝟐. 𝐁𝐨𝐨𝐭𝐥𝐨𝐚𝐝𝐞𝐫 (𝐆𝐑𝐔𝐁): GRand Unified Bootloader (GRUB) is a common choice. It allows you to choose your Linux kernel and initial RAM disk (initrd) if needed. 
 
𝟑. 𝐊𝐞𝐫𝐧𝐞𝐥 𝐈𝐧𝐢𝐭𝐢𝐚𝐥𝐢𝐳𝐚𝐭𝐢𝐨𝐧: ... 
 
🧐𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰𝐞𝐫: Wait, can you elaborate more about Bootloader ? may be about its stages? 
 
😅𝐌𝐞: ok sure.. 
 
𝟐.𝟏 𝐁𝐨𝐨𝐭𝐥𝐨𝐚𝐝𝐞𝐫 𝐒𝐭𝐚𝐠𝐞𝐬: 
 
Stage 1 (MBR/PBR): The first stage of GRUB is usually found in the Master Boot Record (MBR) or the Partition Boot Record (PBR) of the boot device. It's responsible for locating the next stage. 
 
Stage 1.5 (Optional): In complex storage setups, a Stage 1.5 may come into play, understanding file systems and facilitating the loading of Stage 2 
 
Stage 2 (core.img): The critical second stage of GRUB, stored in /boot/grub/, takes over. It presents the GRUB menu, loads the Linux kernel, and initializes the boot process. 
 
𝟑. 𝐊𝐞𝐫𝐧𝐞𝐥 𝐈𝐧𝐢𝐭𝐢𝐚𝐥𝐢𝐳𝐚𝐭𝐢𝐨𝐧: Post-GRUB, the Linux kernel (vmlinuz) and possibly the initial RAM disk (initrd.img) are loaded into memory. The kernel's start_kernel() function is executed, beginning the kernel's internal initialization process. 
 
𝟒. 𝐈𝐧𝐢𝐭-𝐫𝐚𝐦-𝐟𝐬, 𝐈𝐧𝐢𝐭 𝐏𝐫𝐨𝐜𝐞𝐬𝐬, 𝐚𝐧𝐝 𝐁𝐞𝐲𝐨𝐧𝐝: The process continues as we discussed earlier, with initramfs, the init process (often systemd in modern Ubuntu), and user-space services.

**Comment by PranabNandy** [comment]
👩‍💻𝐏𝐚𝐬𝐭 𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰 𝐄𝐱𝐩𝐞𝐫𝐢𝐞𝐧𝐜𝐞: 
The interview started with a discussion on my previous projects and experience with embedded systems. We delved into some technical questions and problem-solving scenarios. 

One question that stood out was related to optimizing code for an embedded system with limited resources(Always limited 😏 ).

💡 𝐈𝐧𝐭𝐞𝐫𝐯𝐢𝐞𝐰 𝐐𝐮𝐞𝐬𝐭𝐢𝐨𝐧:

Question: "So How would you optimize following C code snippet for an embedded system with limited RAM and Flash memory?"
I shared the following approach: 

[hashtag#𝙞𝙣𝙘𝙡𝙪𝙙𝙚](https://www.linkedin.com/feed/hashtag/?keywords=include&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7106881550138449921) <𝙨𝙩𝙙𝙞𝙤.𝙝>

/* 𝘿𝙚𝙛𝙞𝙣𝙚 𝙖 𝙘𝙪𝙨𝙩𝙤𝙢 𝙙𝙖𝙩𝙖 𝙨𝙩𝙧𝙪𝙘𝙩𝙪𝙧𝙚 𝙩𝙤 𝙤𝙥𝙩𝙞𝙢𝙞𝙯𝙚 𝙢𝙚𝙢𝙤𝙧𝙮 𝙪𝙨𝙖𝙜𝙚 */
𝙩𝙮𝙥𝙚𝙙𝙚𝙛 𝙨𝙩𝙧𝙪𝙘𝙩 {
 𝙞𝙣𝙩 𝙭;
 𝙛𝙡𝙤𝙖𝙩 𝙮;
 𝙘𝙝𝙖𝙧 𝙯;
} 𝘾𝙤𝙢𝙥𝙖𝙘𝙩𝘿𝙖𝙩𝙖;

𝙞𝙣𝙩 𝙢𝙖𝙞𝙣() {
 // 𝘾𝙧𝙚𝙖𝙩𝙚 𝙖𝙣 𝙖𝙧𝙧𝙖𝙮 𝙤𝙛 𝘾𝙤𝙢𝙥𝙖𝙘𝙩𝘿𝙖𝙩𝙖
 𝘾𝙤𝙢𝙥𝙖𝙘𝙩𝘿𝙖𝙩𝙖 𝙙𝙖𝙩𝙖_𝙖𝙧𝙧𝙖𝙮[100];

 // 𝙄𝙣𝙞𝙩𝙞𝙖𝙡𝙞𝙯𝙚 𝙙𝙖𝙩𝙖_𝙖𝙧𝙧𝙖𝙮 𝙚𝙡𝙚𝙢𝙚𝙣𝙩𝙨
 𝙛𝙤𝙧 (𝙞𝙣𝙩 𝙞 = 0; 𝙞 < 100; ++𝙞) {
 𝙙𝙖𝙩𝙖_𝙖𝙧𝙧𝙖𝙮[𝙞].𝙭 = 𝙞;
 𝙙𝙖𝙩𝙖_𝙖𝙧𝙧𝙖𝙮[𝙞].𝙮 = 2.5 * 𝙞;
 𝙙𝙖𝙩𝙖_𝙖𝙧𝙧𝙖𝙮[𝙞].𝙯 = '𝘼' + (𝙞 % 26);
 }

 // 𝘼𝙘𝙘𝙚𝙨𝙨 𝙖𝙣𝙙 𝙢𝙖𝙣𝙞𝙥𝙪𝙡𝙖𝙩𝙚 𝙙𝙖𝙩𝙖_𝙖𝙧𝙧𝙖𝙮 𝙖𝙨 𝙣𝙚𝙚𝙙𝙚𝙙

 𝙧𝙚𝙩𝙪𝙧𝙣 0;
}

ℚ𝕦𝕚𝕔𝕜 𝕢𝕦𝕖𝕤𝕥𝕚𝕠𝕟 ~ ℍ𝕠𝕨 𝕨𝕠𝕦𝕝𝕕 𝕪𝕠𝕦 𝕙𝕒𝕧𝕖 𝕠𝕡𝕥𝕚𝕞𝕚𝕫𝕖𝕕 𝕥𝕙𝕖 𝕔𝕠𝕕𝕖?

🤓 In the code, I emphasized the importance of data structure design for memory optimization. By grouping related data fields into a custom structure (CompactData), we minimize padding and reduce memory overhead. This approach helps maximize the efficient use of RAM and Flash memory, a crucial consideration in embedded systems.

The interviewer appreciated the solution and we discussed further optimizations like using bitfields, minimizing global variables, and avoiding dynamic memory allocation.

Remember, interview questions like these often focus on your problem-solving skills and understanding of embedded systems constraints. Tailoring your solution to the specific requirements of the system is key.
Hope this helps fellow embedded enthusiasts! Good luck with your interviews! 💪

**Comment by PranabNandy** [comment]
𝗪𝗵𝗮𝘁 𝗶𝘀 𝗨𝗦𝗕 𝘁𝗼 𝗧𝗧𝗟?

USB to TTL is a small device that allows your computer’s USB port to communicate with other devices using a simpler communication method called TTL (Transistor-Transistor Logic). This is essential when working with devices like microcontrollers, which use serial communication instead of USB.

𝗛𝗼𝘄 𝗗𝗼𝗲𝘀 𝗜𝘁 𝗪𝗼𝗿𝗸?
𝗦𝗲𝗿𝗶𝗮𝗹 𝗖𝗼𝗺𝗺𝘂𝗻𝗶𝗰𝗮𝘁𝗶𝗼𝗻: Many small electronic devices communicate using a method called serial communication. In serial communication, data is sent one bit at a time over a single wire, which is simple but effective for many applications.
𝗧𝗿𝗮𝗻𝘀𝗹𝗮𝘁𝗶𝗼𝗻 𝗼𝗳 𝗦𝗶𝗴𝗻𝗮𝗹𝘀: Your computer sends data using USB, but most microcontrollers and sensors speak “serial.” The USB to TTL converter translates the USB signals from your computer into TTL-level serial signals. This allows your computer to send and receive data to and from the microcontroller.
𝗩𝗼𝗹𝘁𝗮𝗴𝗲 𝗟𝗲𝘃𝗲𝗹𝘀: TTL signals usually operate at 3.3V or 5V, which are safe for small electronic components. The converter ensures that the voltage levels are correct so that your devices don’t get damaged.

𝗘𝘅𝗮𝗺𝗽𝗹𝗲:
Imagine you’re working on a project where your microcontroller needs to send sensor data to your computer. Here’s how USB to TTL makes it happen:
𝗦𝗲𝗻𝘀𝗼𝗿 𝗗𝗮𝘁𝗮: The microcontroller collects data from a temperature sensor. This data is in the form of serial data—bits of information sent one after the other.
𝗗𝗮𝘁𝗮 𝗧𝗿𝗮𝗻𝘀𝗺𝗶𝘀𝘀𝗶𝗼𝗻: The microcontroller sends this serial data out through its serial port, which operates at TTL voltage levels.
𝗨𝗦𝗕 𝘁𝗼 𝗧𝗧𝗟 𝗖𝗼𝗻𝘃𝗲𝗿𝘁𝗲𝗿: You connect the microcontroller to your computer using a USB to TTL converter. The converter translates the serial data into a form that your computer can understand through its USB port.
𝗗𝗮𝘁𝗮 𝗼𝗻 𝗬𝗼𝘂𝗿 𝗖𝗼𝗺𝗽𝘂𝘁𝗲𝗿: Your computer receives the sensor data, which you can then process, log, or display in real-time.

𝗪𝗵𝘆 𝗜𝘁’𝘀 𝗜𝗺𝗽𝗼𝗿𝘁𝗮𝗻𝘁
USB to TTL converters are essential tools for anyone working with embedded systems, IoT devices, or DIY electronics. They provide a reliable way to connect your computer to various devices, enabling you to seamlessly program, debug, and interact with your projects.

**Comment by PranabNandy** [comment]
SPI Communication ☄ 

SPI (Serial Peripheral Interface) is a key protocol in embedded systems, allowing microcontrollers to communicate with devices like sensors and displays. Here’s a quick overview:
Master-Slave Setup: One master device controls the communication, and the others are slaves.
Four Main Lines:
➡ MOSI: Master Out, Slave In
➡ MISO: Master In, Slave Out
➡ SCLK: Serial Clock
➡ SS: Slave Select
Full-Duplex: Data can be sent and received at the same time.
Flexible Clock Settings: SPI can be configured for different devices with clock polarity and phase settings.
SPI is fast and efficient, making it essential for many IoT and embedded applications.
Learn more at
https://media.licdn.com/dms/image/v2/D4D22AQHkCt1rGSDFwA/feedshare-shrink_2048_1536/feedshare-shrink_2048_1536/0/1722947307172?e=1727308800&v=beta&t=WSV76hCmJCq-5LeWj8kFTJN-jfqoRFFcJlPepEObiy8

**Comment by PranabNandy** [comment]
UART Communication 
UART (Universal Asynchronous Receiver/Transmitter) is a fundamental serial communication protocol widely used in embedded systems. It enables devices to exchange data without a shared clock, making it simple and efficient for short-distance communication.

🔹 Key Features:
Asynchronous: No shared clock needed.
Data Packets: Start bit, data bits, optional parity bit, stop bits.
Baud Rate: Common rates include 9600, 19200, 115200 bps.

🔹 Pros:
Easy to implement.
Full-duplex capability.
Optional error detection with parity bits.

🔹 Cons:
Best for short distances.
Limited speed compared to protocols like SPI/I2C.
Ideal for microcontroller communication, peripheral interfaces, and legacy serial devices. Understanding UART is crucial for designing and debugging embedded systems.

**Comment by PranabNandy** [comment]
Understanding System Calls vs. Library Calls

System Call:

A system call is a function provided by the operating system (OS) that allows programs to request services from the kernel, such as accessing hardware or managing system resources. System calls are essential for performing tasks like reading from a file or handling network communication, as they provide a controlled interface to interact with hardware without needing to manage the details directly.

Library Function:

Library functions, on the other hand, are predefined functions that operate in user space. These functions are part of libraries (like the C Standard Library) and help with a wide range of tasks, from string manipulation to complex mathematical operations, making coding more efficient and modular.

Key Differences:

1. Communication Levels:

Library Call: Involves direct communication between the application and the library.

System Call: Involves three levels of communication—first, the application calls a wrapper function, which then triggers a software interrupt to interact with the kernel, where the system call is executed.

2. Argument Passing:

Library Call: Arguments are typically passed via the stack.

System Call: Arguments are often passed through CPU registers, providing faster access for the kernel.

3. Invocation Method:

Library Call: Functions are invoked by their name or memory address.

System Call: Invoked using a unique ID and generally involves a software interrupt.

4. Wrapper Routine:

Library Call: Called directly without the need for a wrapper.

System Call: Usually involves a wrapper routine to switch from user mode to kernel mode.

5. User Friendliness:

Library Call: Abstracts system complexities, making them user-friendly.

System Call: Provides essential functions while abstracting hardware details, but involves more complexity due to the kernel interaction.

**Comment by PranabNandy** [comment]
![Screenshot from 2024-08-26 22-34-58](https://github.com/user-attachments/assets/06280df5-5a02-475b-b14b-370ec482ac61)
![Screenshot from 2024-08-26 22-35-25](https://github.com/user-attachments/assets/d20bb200-6430-4c3d-93b5-41c8bea07eec)

**Comment by C-mmon** [comment]
GG
